src/Repartition.chpl

```diff
@@ -24,6 +27,495 @@ module Repartition
+    // not just the portion on that locale).
+    var firstStringSizes: [0..#numLocales] int = -1;
+    var lastStringSizes: [0..#numLocales] int = -1;
+
+    // Number of bytes of the first/last string that actually reside on each locale.
+    // This may be less than the full string size if the string is split across locales.
+    var firstStringNumBytesEachLocale: [0..#numLocales] int = -1;
+    var lastStringNumBytesEachLocale: [0..#numLocales] int = -1;
+
+    // Now, on each locale, we examine the portion of the offsets that live here to
+    // figure out which strings intersect each locale's local byte chunk.
+    coforall loc in Locales do on loc {
+      // Upper bound (exclusive) for strings that start on this locale
+      const myLastOffsetEnd = lastOffsetEndPerLoc[here.id];
+
+      // Local slice of the offsets that actually reside on this locale
+      ref myOffsets = offsets.localSlice[offsets.localSubdomain()];
+
+      // Local copies of global first/last byte indices per locale
+      const myFirstIdxPerLoc = firstIdxPerLoc;
+      const myLastIdxPerLoc = lastIdxPerLoc;
+
+      if !myOffsets.isEmpty() {
+
+        // For *every* locale i, check whether that locale's byte range overlaps
+        // any string whose starting offset is in myOffsets (i.e., whose segment
+        // begins on this locale).
+        forall i in 0..#numLocales {
+          // ----- FIRST STRING FOR LOCALE i -----
+          // If locale i has bytes, and its first byte index falls within the
+          // range of strings that start on this locale, then find which string
+          // contains that starting byte.
+          if myFirstIdxPerLoc[i] >= myOffsets.first && myFirstIdxPerLoc[i] < myLastOffsetEnd {
+            // Find the greatest offset in myOffsets that is <= myFirstIdxPerLoc[i]
+            // binarySearch returns (found?, index)
+            var result = binarySearch(myOffsets, myFirstIdxPerLoc[i]);
+            const idx = if result[0] then result[1] else result[1] - 1;
+
+            // Record the global string index for the first string on locale i
+            firstStringIndices[i] = idx;
+            firstStringStartOffsets[i] = myOffsets[idx];
+
+            // Determine the end of that string: next offset if available,
+            // otherwise the lastOffsetEnd for this locale.
+            const endStr = if idx < myOffsets.domain.last
+                           then myOffsets[idx + 1]
+                           else myLastOffsetEnd;
+
+            // Full size of the string in bytes
+            const size = endStr - myOffsets[idx];
+            firstStringSizes[i] = size;
+
+            // Portion of this string that resides on locale i's byte range
+            // (truncate end to not go past that locale’s last byte + 1)
+            const endStrThisLoc = min(endStr, myLastIdxPerLoc[i] + 1);
+            firstStringNumBytesEachLocale[i] = endStrThisLoc - myFirstIdxPerLoc[i];
```

Comment:
And another one.

These could be benign write-after-write hazards, though I still wanted to check.